python/cudf_polars/cudf_polars/experimental/expressions.py

```diff
@@ -0,0 +1,484 @@
+)
+
+
+_SUPPORTED_AGGS = ("count", "min", "max", "sum", "mean", "n_unique")
+
+
+class _Placeholder(IR):
+    """
+    Placeholder leaf node.
+
+    Notes
+    -----
+    This is used as the required IR for ``Literal``
+    expressions in ``_decompose``.
+    """
+
+    __slots__ = ()
+    _non_child = ("schema",)
+
+    def __init__(self, schema: Schema):
+        self.schema = schema
+        self._non_child_args = ()
+        self.children = ()
+
+    @classmethod
+    def do_evaluate(cls) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        return DataFrame([])
+
+
+def _add_select_ir(
+    exprs: Sequence[Expr],
+    input_ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    *,
+    names: Generator[str, None, None],
+    repartition: bool = False,
+) -> tuple[list[Col], Repartition | Select, MutableMapping[IR, PartitionInfo]]:
+    """
+    Add a new Select node.
+
+    Parameters
+    ----------
+    exprs
+        Expressions to select.
+    input_ir
+        The input IR node to select from.
+    partition_info
+        A mapping from all unique IR nodes to the
+        associated partitioning information.
+    names
+        Generator of unique names for temporaries.
+    repartition
+        Whether to add a Repartition node after the
+        new selection.
+
+    Returns
+    -------
+    columns
+        New expressions to use with ``new_ir``.
+    new_ir
+        The new IR node that can be evaluated by ``columns``.
+    partition_info
+        A mapping from unique nodes in the new graph to associated
+        partitioning information.
+
+    See Also
+    --------
+    _decompose_agg_node
+
+    Notes
+    -----
+    This function is called by ``_decompose_agg_node`` to decompose
+    Agg operations into multiple IR nodes.
+    """
+    output_names = [next(names) for _ in range(len(exprs))]
+    named_exprs = [
+        NamedExpr(name, expr) for name, expr in zip(output_names, exprs, strict=True)
+    ]
+    new_ir: Repartition | Select = Select(
+        {ne.name: ne.value.dtype for ne in named_exprs},
+        named_exprs,
+        True,  # noqa: FBT003
+        input_ir,
+    )
+    partition_info[new_ir] = PartitionInfo(count=partition_info[input_ir].count)
+
+    # Optionally collapse into one output partition
+    if repartition:
+        new_ir = Repartition(new_ir.schema, new_ir)
+        partition_info[new_ir] = PartitionInfo(count=1)
+
+    columns = [Col(ne.value.dtype, ne.name) for ne in named_exprs]
+    return columns, new_ir, partition_info
+
+
+def _maybe_shuffle(
+    child: Expr,
+    input_ir: IR,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    config_options: ConfigOptions,
+    *,
+    names: Generator[str, None, None],
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Inject shuffle (if necessary)
+    # Used by `_decompose_agg_node`
+    pi = partition_info[input_ir]
+    if pi.count > 1 and [ne.value for ne in pi.partitioned_on] != [child]:
```

Comment:
Perhaps flip the condition?
```
    if pi.count == 1 and len(pi.partitioned_on) == 1 and child == pi.partitioned_on[0].value:
         return input_ir, partition_info
    
```
And then in the else fall-through do the shuffle.

Reply:
I'll flip it, but we want to avoid the shuffle for multi-partition data that is already shuffled as well. So:

```
    if pi.count == 1 or [ne.value for ne in pi.partitioned_on] == [child]:
        return input_ir, partition_info
    else:
        from cudf_polars.experimental.shuffle import Shuffle
        ...
```